model.py

```diff
@@ -1,103 +1,1394 @@
+        other branches
+
+        vert_gap: gap between levels of hierarchy
+
+        vert_loc: vertical location of root
+
+        xcenter: horizontal location of root
+        '''
+        if not nx.is_tree(G):
+            fail_msg = 'cannot use hierarchy_pos on a graph that is not a tree'
+            raise TypeError(fail_msg)
+
+        if root is None:
+            if isinstance(G, nx.DiGraph):
+                # Allows back compatibility with nx version 1.11
+                root = next(iter(nx.topological_sort(G)))
+            else:
+                import random
+                root = random.choice(list(G.nodes))
+
+        return self._hierarchy_pos(G, root, width, vert_gap, vert_loc, xcenter)
+
+    def _hierarchy_pos(self, G, root,
+                       width=1., vert_gap=0.2, vert_loc=0, xcenter=0.5,
+                       pos=None, parent=None):
+        '''
+        see hierarchy_pos docstring for most arguments
+
+        pos: a dict saying where all nodes go if they have been assigned
+        parent: parent of this branch. - only affects it if non-directed
+
+        '''
+
+        if pos is None:
+            pos = {root: (xcenter, vert_loc)}
+        else:
+            pos[root] = (xcenter, vert_loc)
+
+        children = list(G.neighbors(root))
+
+        if not isinstance(G, nx.DiGraph) and parent is not None:
+            children.remove(parent)
+
+        if len(children) != 0:
+            dx = width/len(children)
+            nextx = xcenter - width/2 - dx/2
+            for child in children:
+                nextx += dx
+                pos = self._hierarchy_pos(
+                    G, child,
+                    width=dx, vert_gap=vert_gap,
+                    vert_loc=vert_loc-vert_gap, xcenter=nextx,
+                    pos=pos, parent=root
+                )
+
+        return pos
+
+
+# Subclasses.
+class Eclipse(Model):
+    '''Subclass of Model, specifically for storing a single eclipse. Lightcurve
+    data is stored on this level.
+
+    Inputs:
+    -------
+      lightcurve; Lightcurve:
+        A Lightcurve object, containing data
+      iscomplex; bool:
+        Do we use the complex model?
+      label; str:
+        A label to apply to the node. Mostly used when searching trees.
+      parameter_objects; list(Param), or Param:
+        The parameter objects that correspond to this node. Single Param is
+        also accepted.
+      parent; Model, optional:
+        The parent of this node.
+      children; list(Model), or Model:
+        The children of this node. Single Model is also accepted
+    '''
+
+    # Define this subclasses parameters
+    node_par_names = (
+        'dFlux', 'sFlux', 'ulimb', 'rdisc',
+        'scale', 'az', 'fis', 'dexp', 'phi0'
+    )
+
+    def __init__(self, lightcurve, iscomplex, *args, **kwargs):
+        # If we're a complex eclipse, add the complex parameter names
+        if iscomplex:
```

Comment:
I'd still prefer two classes, one for a ```SimpleEclipse``` and one for a ```ComplexEclipse```, with different ```node_par_names``` for each

This seems fragile to me; for example, we want to know that ```Eclipse.node_par_names``` is correct, even before an instance of ```Eclipse``` is defined, but you can't do that, because ```node_par_names``` is being changed on initialisation

Reply:
That's a very good point. I think changing Eclipse to SimpleEclipse, and subclassing ComplexEclipse from that would be easiest to understand in 6 months when I've forgotten how I wrote this...